lookup_estim.go

```diff
@@ -0,0 +1,302 @@
+	// don't terminate the process early because we can't find any closer peers.
+	OptProvSetThresholdStrictness = 0.1
+
+	// OptProvReturnRatio corresponds to how many ADD_PROVIDER RPCs must have completed (regardless of success)
+	// before we return to the user. The ratio of 0.75 equals 15 RPC as it is based on the Kademlia bucket size.
+	OptProvReturnRatio = 0.75
+)
+
+type addProviderRPCState int
+
+const (
+	Sent addProviderRPCState = iota + 1
+	Success
+	Failure
+)
+
+type estimatorState struct {
+	// context for all ADD_PROVIDER RPCs
+	putCtx context.Context
+
+	// reference to the DHT
+	dht *IpfsDHT
+
+	// the most recent network size estimation
+	networkSize float64
+
+	// a channel indicating when an ADD_PROVIDER RPC completed (successful or not)
+	doneChan chan struct{}
+
+	// tracks which peers we have stored the provider records with
+	peerStatesLk sync.RWMutex
+	peerStates   map[peer.ID]addProviderRPCState
+
+	// the key to provide
+	key string
+
+	// the key to provide transformed into the Kademlia key space
+	ksKey ks.Key
+
+	// distance threshold for individual peers. If peers are closer than this number we store
+	// the provider records right away.
+	individualThreshold float64
+
+	// distance threshold for the set of bucketSize closest peers. If the average distance of the bucketSize
+	// closest peers is below this number we stop the DHT walk and store the remaining provider records.
+	// "remaining" because we have likely already stored some on peers that were below the individualThreshold.
+	setThreshold float64
+
+	// number of completed (regardless of success) ADD_PROVIDER RPCs before we return control back to the user.
+	returnThreshold int
+
+	// putProvDone counts the ADD_PROVIDER RPCs that have completed (successful and unsuccessful)
+	putProvDone atomic.Int32
+}
+
+func (dht *IpfsDHT) newEstimatorState(ctx context.Context, key string) (*estimatorState, error) {
+	// get network size and err out if there is no reasonable estimate
+	networkSize, err := dht.nsEstimator.NetworkSize()
+	if err != nil {
+		return nil, err
+	}
+
+	individualThreshold := mathext.GammaIncRegInv(float64(dht.bucketSize), 1-OptProvIndividualThresholdCertainty) / networkSize
+	setThreshold := mathext.GammaIncRegInv(float64(dht.bucketSize)/2.0+1, 1-OptProvSetThresholdStrictness) / networkSize
+	returnThreshold := int(math.Ceil(float64(dht.bucketSize) * OptProvReturnRatio))
+
+	return &estimatorState{
+		putCtx:              ctx,
+		dht:                 dht,
+		key:                 key,
+		doneChan:            make(chan struct{}, returnThreshold), // buffered channel to not miss events
+		ksKey:               ks.XORKeySpace.Key([]byte(key)),
+		networkSize:         networkSize,
+		peerStates:          map[peer.ID]addProviderRPCState{},
+		individualThreshold: individualThreshold,
+		setThreshold:        setThreshold,
+		returnThreshold:     returnThreshold,
+		putProvDone:         atomic.Int32{},
+	}, nil
+}
```

Comment:
AFAIT `estimatorState` takes two inputs:
1. The key
2. The size of the network

That the optimistic client code! That not related to estimating the network. I would rename this to `optimisticState` to make it clear where the two different features split.

Reply:
Sounds good, I've renamed the struct :+1: